README
fwsnort   (Firewall Snort)
Version:  0.6.4
Author:   Michael Rash <mbr@cipherdyne.org>
Website:  http://www.cipherdyne.org/fwsnort/

=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
INSTALLATION:

    (See the INSTALL file in the source directory.)

=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
**IMPORTANT**:

fwsnort requires the iptables string match module in order to be able to
detect application layer attacks.  In addition, fwsnort extends iptables by
adding a "--hex-string" option with a patch against
iptables/extensions/libipt_string.c.  This patch has been accepted by the
iptables maintainers and is a part of iptables as of the 1.2.9 release.
If you are running a version of iptables that is less than 1.2.9, then you
will need to apply the patch yourself (the patch is against
iptables-1.2.7a).  The basic steps to do this are:

-download iptables-1.2.7a.tar.bz2 from http://www.netfilter.org
-tar xvfj iptables-1.2.7a.tar.bz2
-cp fwsnort-0.1/libipt_string.c.patch iptables-1.2.7a/extensions
-cd iptables-1.2.7a/extensions
-patch libipt_string.c libipt_string.c.patch
-compile and install iptables

=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
DESCRIPTION:

fwsnort is a perl script that translates snort rules into equivalent iptables
rules.  Some snort rule options such as "content-list", and "sameip" have no
direct translation into iptables options so not all snort rules can be
translated.  However approximately 65% of all snort signatures can be
successfully translated through the use of the iptables string match module.
Also, included with the fwsnort sources is a patch for the user space portion
of iptables that adds a new option "--hex-string" which accepts an argument
such as "|5a 4e|".  This allows the content fields in snort rules to be
directly input into iptables rulesets from the command line.  fwsnort parses
the running iptables policy on the machine in order to determine which snort
rules are applicable to the specific policy loaded on the machine.

NOTE: The Netfilter string match extension has not yet been ported to the
2.6 Linux kernel, and so fwsnort is yet not compatible with the 2.6 kernel.

Also, note that string_replace_iptables.patch and
string_replace_iptables.patch are included in the fwsnort sources, but they
are not actually used by fwsnort.  These patches provide the ability to
alter application layer data in a manner similar to the "replace" keyword
in Snort_inline by adding two new iptables command line options
"--replace-string" and "--replace-hex-string".

=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
PLATFORMS:

fwsnort is compatible with iptables only, hence fwsnort will exclusively run
on Linux running a 2.4.x series kernel.

=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
COPYRIGHT:

Copyright (C) 2003 Michael Rash (mbr@cipherdyne.org)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


$Id: README 176 2004-12-19 02:46:43Z mbr $


ChangeLog
fwsnort-1.0.4 (01/22/2008):
    - (Grant Ferley) Submitted patch to exclude loopback interfaces from
      iptables allow rules parsing.  This behavior can be reversed with the
      existing --no-exclude-loopback command line argument.
    - (Grant Ferley) Submitted patch to IPTables::Parse to take into account
      iptables policy output that contains "0" instead of "all" to represent
      any protocol.
    - (Grant Ferley) Submitted patch to IPTables::Parse to set sport and dport
      to '0:0' if the protocol is 'all'.
    - Bugfix to allow negated networks to be specified within iptables allow
      rules or within the fwsnort.conf file.
    - Updated install.pl to set the LC_ALL environmental variable to "C". This
      should fix potential locale problems (this fix was borrowed from the
      fwknop project).

fwsnort-1.0.3 (11/22/2007):
    - Added --include-re-caseless and --exclude-re-caseless options to have
      --include-regex and --exclude-regex options match case insensitively.
    - Major signature update from Bleeding Threats. This update includes a
      large number of new signatures with PCRE statements, with an emphasis on
      detecting SQL injection attacks directed at internal webservers from
      external sources.
    - Added the ability to interpret PCRE statements that include simple
      string matches separated by ".*" and ".+" as multiple iptables string
      matches. The only negative consequence in terms of signature detection
      is that ordering is not preserved; that is, the PCRE "/UNION.+SELECT/"
      would only match a packet that contains "UNION" followed by "SELECT",
      whereas an iptables rule that uses a string match for UNION and a
      separate string match for SELECT would match a packet that contains both
      strings but in reverse. Typically this is not a huge concern, and the
      PCRE translation can be disabled with a new option --no-pcre.
    - Added asn1 keyword to unsupported list.

fwsnort-1.0.2 (08/26/2007):
    - Bugfix to make sure to add in header lengths for depth and offset values
      since the string match extension compares bytes from the start of the
      data link header.

fwsnort-1.0.1 (08/26/2007):
    - Bugfix for ipt_rule_test() function name.
    - Added the ability to automatically resolve command paths if any commands
      cannot be found at the locations specified in the fwsnort.conf file.

fwsnort-1.0 (04/19/2007):
    - Major update to include support for the NFQUEUE and QUEUE targets with
      new command line options --NFQUEUE and --QUEUE.  This changes the
      default LOG target to the NFQUEUE or QUEUE targets instead, and at the
      same time builds a parallel Snort rule set in the
      /etc/fwsnort/snort_rules_queue directory.  Every Snort rule in this
      directory has at least one "content" keyword, which fwsnort uses in the
      resulting iptables policy.  This policy only sends those packets to
      snort_inline via the NFQUEUE or QUEUE target that match a content field
      within some Snort rule.  The end result is that snort_inline should run
      faster because the vast majority of packets (which are not malicious)
      are processed via the Linux kernel without ever having to be sent to
      userspace for analysis.  There is a tradeoff here in terms of attack
      detection; snort_inline does not receive all packets associated with a
      stream, so it cannot detect attacks quite as effectively (snort_inline
      does not have an opportunity to look at reassembled buffers).  However,
      this trade off may be acceptable for large sites where performance is
      more important.
    - Bug fix to remove any existing jump rules from the built-in INPUT,
      OUTPUT, and FORWARD chains before creating a new jump rules.  This
      allows the fwsnort.sh script to be executed multiple times without
      creating a new jump rule into the fwsnort chains for each execution.
    - Added the -X command line argument to allow fwsnort to delete all of
      the fwsnort chains; this emulates the iptables command line argument
      of the same name.
    - Minor output enhancements and bugfixes to give more insight into the
      translation process.  For example, if fwsnort is run in --snort-sid
      mode but is unable to translate the specified signatures, the user is
      notified.  Also, any existing /etc/fwsnort/fwsnort.sh script is not
      archived and erased until fwsnort is actually going to write a new one.
    - Added sid values to iptables comment match string.
    - Bugfix for iptables string match --from and --to values to skip past
      packet headers.  This is an approximation until a new --payload option
      can be added to the string match extension.
    - Added a single iptables rule testing API internally within fwsnort;
      this adds a measure of consistency and removes some duplicate code.
    - Added fwsnort mailing list at SourceForge.

fwsnort-0.9.0 (03/22/2007):
    - Added support for multiple content matches since this is supported by
      iptables.  This made a 10% increase in the fwsnort translation rate -
      about 60% of all Snort-2.3.3 rules can be translated now.
    - Added emulation for distance and within from previous content match
      based on --from and --to (string match extension) and the length of
      the previous pattern.
    - Added the ability to include the Snort "msg", "classtype", "reference",
      "priority", and "rev" fields in each iptables rule with the comment
      match.  This can be disabled with a new command line argument
      --no-ipt-comments.  The fwsnort version is also included within this
      string.
    - Added the ability to include the iptables rule number for each rule in
      the fwsnort chains.  This is useful to easily know which iptables rule
      is being triggered by network traffic (so it can be disabled if
      necessary).  This can be disabled with --no-ipt-rule-nums.

snort_opts.pl
#!/usr/bin/perl -w
#
########################################################################
#
# File: snort_opts.pl
#
# Purpose: To parse snort rules and display a listing of snort fields
#          along with how many snort rules in which each field is
#          found.
#
########################################################################
#
# $Id: snort_opts.pl 290 2007-02-15 04:18:19Z mbr $
#

my %options = (
    'flow'         => 0,
    'flowbits'     => 0,
    'msg'          => 0,
    'logto'        => 0,
    'ttl'          => 0,
    'tos'          => 0,
    'id'           => 0,
    'ipopts'       => 0,
    'fragbits'     => 0,
    'dsize'        => 0,
    'flags'        => 0,
    'seq'          => 0,
    'ack'          => 0,
    'itype'        => 0,
    'icode'        => 0,
    'icmp_id'      => 0,
    'icmp_seq'     => 0,
    'content'      => 0,
    'uricontent'   => 0,
    'content-list' => 0,
    'offset'       => 0,
    'depth'        => 0,
    'nocase'       => 0,
    'session'      => 0,
    'rpc'          => 0,
    'resp'         => 0,
    'react'        => 0,
    'reference'    => 0,
    'sid'          => 0,
    'rev'          => 0,
    'classtype'    => 0,
    'priority'     => 0,
    'tag'          => 0,
    'ip_proto'     => 0,
    'sameip'       => 0,
    'stateless'    => 0,
    'regex'        => 0,
    'distance'     => 0,
    'within'       => 0,
    'byte_jump'    => 0,
    'byte_test'    => 0,
    'pcre'         => 0
);


my $dir   = 'snort_rules';
my $total_rules = 0;

opendir D, $dir or die " ** Could not open $dir: $!";
my @rfiles = readdir D;
closedir D;

print " .. Calculating snort rule keyword percentages:\n";
for my $rfile (@rfiles) {
    next unless $rfile =~ /\.rules/;
    open R, "< $dir/$rfile" or die $!;
    my @lines = <R>;
    close R;

    for my $line (@lines) {
        chomp $line;
        next unless $line =~ /\S/;
        next if $line =~ /^#/;
        $total_rules++;
        if ($line =~ /^\s*alert/) {
            for my $opt (keys %options) {
                if ($line =~ /\s$opt[:;]/) {
                    $options{$opt}++;
                } elsif ($line =~ /\($opt[:;]/) {
                    $options{$opt}++;
                } elsif ($line =~ /;$opt[:;]/) {
                    $options{$opt}++;
                }
            }
        }
    }
}
for my $opt (sort {$options{$b} <=> $options{$a}} keys %options) {
    printf("%13s %13s", $opt, "$options{$opt}/$total_rules  ");
    print sprintf("%.1f", $options{$opt} / $total_rules * 100) . "%\n";
}

exit 0;


snortspoof.pl
#!/usr/bin/perl -w
#
###############################################################################
#
# File: snortspoof.pl
#
# Purpose: To parse rules from the Snort rule set and spoof them at a target
#          IP from arbitrary source addresses.  This is similar to the
#          technique employed by the Stick and Snot projects.  Snortspoof.pl
#          is distributed with the fwsnort project
#          (http://www.cipherdyne.org/fwsnort/).
#
# Author: Michael Rash <mbr@cipherdyne.org>
#
# Copyright (C) 2003-2007 Michael Rash (mbr@cipherdyne.org)
#
# License (GNU Public License):
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
#    USA
#
###############################################################################
#
# $Id: snortspoof.pl 335 2007-03-12 02:07:49Z mbr $
#

require Net::RawIP;
use strict;

my $file       = $ARGV[0] || '';
my $spoof_addr = $ARGV[1] || '';
my $dst_addr   = $ARGV[2] || '';

die "$0 <rules file> <spoof IP> <dst IP>"
    unless $file and $spoof_addr and $dst_addr;

# alert udp $EXTERNAL_NET any -> $HOME_NET 635 (msg:"EXPLOIT x86 Linux mountd
# overflow"; content:"^|B0 02 89 06 FE C8 89|F|04 B0 06 89|F";
# reference:bugtraq,121; reference:cve,1999-0002; classtype:attempted-admin;
# sid:315; rev:6;)
my $sig_sent = 0;
open F, "< $file" or die "[*] Could not open $file: $!";
SIG: while (<F>) {
    my $content = '';
    my $conv_content = '';
    my $hex_mode = 0;
    my $proto = '';
    my $spt = 10000;
    my $dpt = 10000;

    ### make sure it is an inbound sig
    if (/^\s*alert\s+(tcp|udp)\s+\S+\s+(\S+)\s+\S+
            \s+(\$HOME_NET|any)\s+(\S+)\s/x) {
        $proto = $1;
        my $spt_tmp = $2;
        my $dpt_tmp = $4;

        ### can't handle multiple content fields yet
        next SIG if /content:.*\s*content\:/;

        $content = $1 if /\s*content\:\"(.*?)\"\;/;
        next SIG unless $content;

        if ($spt_tmp =~ /(\d+)/) {
            $spt = $1;
        } elsif ($spt_tmp ne 'any') {
            next SIG;
        }
        if ($dpt_tmp =~ /(\d+)/) {
            $dpt = $1;
        } elsif ($dpt_tmp ne 'any') {
            next SIG;
        }

        my @chars = split //, $content;
        for (my $i=0; $i<=$#chars; $i++) {
            if ($chars[$i] eq '|') {
                $hex_mode == 0 ? ($hex_mode = 1) : ($hex_mode = 0);
                next;
            }
            if ($hex_mode) {
                next if $chars[$i] eq ' ';
                $conv_content .= sprintf("%c",
                        hex($chars[$i] . $chars[$i+1]));
                $i++;
            } else {
                $conv_content .= $chars[$i];
            }
        }
        my $rawpkt = '';
        if ($proto eq 'tcp') {
            $rawpkt = new Net::RawIP({'ip' => {
                saddr => $spoof_addr, daddr => $dst_addr},
                'tcp' => { source => $spt, dest => $dpt, 'ack' => 1,
                data => $conv_content}})
                    or die "[*] Could not get Net::RawIP object: $!";
        } else {
            $rawpkt = new Net::RawIP({'ip' => {
                saddr => $spoof_addr, daddr => $dst_addr},
                'udp' => { source => $spt, dest => $dpt,
                data => $conv_content}})
                    or die "[*] Could not get Net::RawIP object: $!";
        }
        $rawpkt->send();
        $sig_sent++;
    }
}
print "[+] $file, $sig_sent attacks sent.\n";
close F;
exit 0;


install.pl
#!/usr/bin/perl -w
#
#######################################################################
#
# File: install.pl
#
# Purpose: This is the installation script for fwsnort.
#
# Copyright (C) 2003-2007 Michael Rash (mbr@cipherdyne.org)
#
# License (GNU Public License):
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
#    USA
#
# TODO:
#   - Write the uninstall() routine. :)
#
#######################################################################
#
# $Id: install.pl 417 2008-01-15 04:39:14Z mbr $
#

use Cwd;
use IO::Socket;
use File::Copy;
use File::Path;
use Getopt::Long;
use strict;

#========================= config ========================
my $sbin_dir    = '/usr/sbin';
my $lib_dir     = '/usr/lib/fwsnort';
my $fwsnort_dir = '/etc/fwsnort';
my $rules_dir   = "${fwsnort_dir}/snort_rules";

### Snort.org no longer allows auto downloads of signatures
my $bleeding_snort_website = 'www.bleedingsnort.com';

### system binaries
my $perlCmd = '/usr/bin/perl';
my $makeCmd = '/usr/bin/make';
my $wgetCmd = '/usr/bin/wget';
my $gzipCmd = '/bin/gzip';
my $tarCmd  = '/bin/tar';
#======================= end config ======================

### map perl modules to versions
my %required_perl_modules = (
    'Net::IPv4Addr' => {
        'force-install' => 0,
        'mod-dir' => 'Net-IPv4Addr'
    },
    'IPTables::Parse' => {
        'force-install' => 1,
        'mod-dir' => 'IPTables-Parse'
    }
);

### establish some defaults
my $uninstall = 0;
my $cmdline_force_install = 0;
my $force_mod_re = '';
my $exclude_mod_re = '';
my $help = 0;
my $locale = 'C';  ### default LC_ALL env variable
my $no_locale = 0;

my $src_dir = getcwd() or die "[*] Could not get current working directory.";

my %cmds = (
    'perl' => $perlCmd,
    'make' => $makeCmd,
    'gzip' => $gzipCmd,
    'wget' => $wgetCmd,
    'tar'  => $tarCmd
);

### make Getopts case sensitive
Getopt::Long::Configure('no_ignore_case');

&usage(1) unless (GetOptions(
    'force-mod-install' => \$cmdline_force_install,  ### force install of all modules
    'Force-mod-regex=s' => \$force_mod_re, ### force specific mod install with regex
    'Exclude-mod-regex=s' => \$exclude_mod_re, ### exclude a particular perl module
    'uninstall' => \$uninstall, ### uninstall fwsnort
    'LC_ALL=s'  => \$locale,
    'no-LC_ALL' => \$no_locale,
    'help'      => \$help
));

&usage(0) if $help;

### set LC_ALL env variable
$ENV{'LC_ALL'} = $locale unless $no_locale;

$force_mod_re = qr|$force_mod_re| if $force_mod_re;
$exclude_mod_re = qr|$exclude_mod_re| if $exclude_mod_re;

### make sure the system binaries are where we think they are.
&check_commands();

### check to make sure we are running as root
$< == 0 && $> == 0 or die "You need to be root (or equivalent UID 0",
    " account) to install/uninstall fwsnort!\n";

if ($uninstall) {
    &uninstall();
} else {
    &install()
}
exit 0;
#===================== end main ===================

sub install() {
    die "[*] You must run install.pl from the fwsnort " .
        "sources directory." unless -e 'fwsnort' and -e 'fwsnort.conf';

    unless (-d $fwsnort_dir) {
        print "[+] mkdir $fwsnort_dir\n";
        mkdir $fwsnort_dir, 0500;
    }
    unless (-d $rules_dir) {
        print "[+] mkdir $rules_dir\n";
        mkdir $rules_dir, 0500;
    }

    ### install perl modules
    for my $module (keys %required_perl_modules) {
        &install_perl_module($module);
    }

    my $local_rules_dir = 'snort_rules';
    if (&query_get_bleeding_snort()) {
        chdir $local_rules_dir or die "[*] Could not chdir $local_rules_dir";
        if (-e 'bleeding-all.rules') {
            move 'bleeding-all.rules', 'bleeding-all.rules.tmp'
                or die "[*] Could not move bleeding-all.rules -> ",
                "bleeding-all.rules.tmp";
        }
        system "$cmds{'wget'} http://$bleeding_snort_website/bleeding-all.rules";
        if (-e 'bleeding-all.rules') {  ### successful download
            unlink 'bleeding-all.rules.tmp';
        } else {
            print "[-] Could not download bleeding-all.rules file.\n";
            if (-e 'bleeding-all.rules.tmp') {
                ### move the original back
                move 'bleeding-all.rules', 'bleeding-all.rules.tmp'
                    or die "[*] Could not move bleeding-all.rules -> ",
                    "bleeding-all.rules.tmp";
            }
        }
        chdir '..';
    }

    opendir D, $local_rules_dir or die "[*] Could not open ",
        "the $local_rules_dir directory: $!";
    my @rfiles = readdir D;
    closedir D;

    print "[+] Copying all rules files to $rules_dir\n";
    for my $rfile (@rfiles) {
        next unless $rfile =~ /\.rules$/;
        print "[+] Installing $rfile\n";
        copy "snort_rules/${rfile}", "${rules_dir}/${rfile}" or
            die "[*] Could not copy snort_rules/${rfile} ",
                "-> ${rules_dir}/${rfile}";
    }

    print "\n";

    ### install the fwsnort.8 man page
    &install_manpage();

    my $preserve_rv = 0;
    if (-e "${fwsnort_dir}/fwsnort.conf") {
        $preserve_rv = &query_preserve_config();
    }

    if ($preserve_rv) {
        &preserve_config();
    } else {
        print "[+] Copying fwsnort.conf -> ${fwsnort_dir}/fwsnort.conf\n";
        copy 'fwsnort.conf', "${fwsnort_dir}/fwsnort.conf";
        chmod 0600, "${fwsnort_dir}/fwsnort.conf";
    }

    print "[+] Copying fwsnort -> ${sbin_dir}/fwsnort\n";
    copy 'fwsnort', "${sbin_dir}/fwsnort";
    chmod 0500, "${sbin_dir}/fwsnort";

    print "\n========================================================\n",
        "\n[+] fwsnort will generate an iptables script located at:\n",
        "    /etc/fwsnort/fwsnort.sh when executed.\n",
        "\n[+] fwsnort has been successfully installed!\n\n";

    return;
}

sub install_perl_module() {
    my $mod_name = shift;

    die '[*] Missing force-install key in required_perl_modules hash.'
        unless defined $required_perl_modules{$mod_name}{'force-install'};
    die '[*] Missing mod-dir key in required_perl_modules hash.'
        unless defined $required_perl_modules{$mod_name}{'mod-dir'};

    if ($exclude_mod_re and $exclude_mod_re =~ /$mod_name/) {
        print "[+] Excluding installation of $mod_name module.\n";
        return;
    }

    my $version = '(NA)';

    my $mod_dir = $required_perl_modules{$mod_name}{'mod-dir'};

    if (-e "$mod_dir/VERSION") {
        open F, "< $mod_dir/VERSION" or
            die "[*] Could not open $mod_dir/VERSION: $!";
        $version = <F>;
        close F;
        chomp $version;
    } else {
        print "[-] Warning: VERSION file does not exist in $mod_dir\n";
    }

    my $install_module = 0;

    if ($required_perl_modules{$mod_name}{'force-install'}
            or $cmdline_force_install) {
        ### install regardless of whether the module may already be
        ### installed
        $install_module = 1;
    } elsif ($force_mod_re and $force_mod_re =~ /$mod_name/) {
        print "[+] Forcing installation of $mod_name module.\n";
        $install_module = 1;
    } else {
        if (has_perl_module($mod_name)) {
            print "[+] Module $mod_name is already installed in the ",
                "system perl tree, skipping.\n";
        } else {
            ### install the module in the /usr/lib/fwknop directory because
            ### it is not already installed.
            $install_module = 1;
        }
    }

    if ($install_module) {
        unless (-d $lib_dir) {
            print "[+] Creating $lib_dir\n";
            mkdir $lib_dir, 0755 or die "[*] Could not mkdir $lib_dir: $!";
        }
        print "[+] Installing the $mod_name $version perl " .
            "module in $lib_dir/\n";
        my $mod_dir = $required_perl_modules{$mod_name}{'mod-dir'};
        chdir $mod_dir or die "[*] Could not chdir to ",
            "$mod_dir: $!";
        unless (-e 'Makefile.PL') {
            die "[*] Your $mod_name source directory appears to be incomplete!\n",
                "    Download the latest sources from ",
                "http://www.cipherdyne.org/\n";
        }
        system "$cmds{'make'} clean" if -e 'Makefile';
        system "$cmds{'perl'} Makefile.PL PREFIX=$lib_dir LIB=$lib_dir";
        system $cmds{'make'};
#        system "$cmds{'make'} test";
        system "$cmds{'make'} install";
        chdir $src_dir or die "[*] Could not chdir $src_dir: $!";

        print "\n\n";
    }
    return;
}

sub has_perl_module() {
    my $module = shift;

    # 5.8.0 has a bug with require Foo::Bar alone in an eval, so an
    # extra statement is a workaround.
    my $file = "$module.pm";
    $file =~ s{::}{/}g;
    eval { require $file };

    return $@ ? 0 : 1;
}

sub uninstall() {
    ### FIXME
    return;
}

sub install_manpage() {
    my $manpage = 'fwsnort.8';
    ### remove old man page
    unlink "/usr/local/man/man8/${manpage}" if
        (-e "/usr/local/man/man8/${manpage}");

    ### default location to put the fwsnort man page, but check with
    ### /etc/man.config
    my $mpath = '/usr/share/man/man8';
    if (-e '/etc/man.config') {
        ### prefer to install $manpage in /usr/local/man/man8 if
        ### this directory is configured in /etc/man.config
        open M, '< /etc/man.config' or
            die "[*] Could not open /etc/man.config: $!";
        my @lines = <M>;
        close M;
        ### prefer the path "/usr/share/man"
        my $found = 0;
        for my $line (@lines) {
            chomp $line;
            if ($line =~ m|^MANPATH\s+/usr/share/man|) {
                $found = 1;
                last;
            }
        }
        ### try to find "/usr/local/man" if we didn't find /usr/share/man
        unless ($found) {
            for my $line (@lines) {
                chomp $line;
                if ($line =~ m|^MANPATH\s+/usr/local/man|) {
                    $mpath = '/usr/local/man/man8';
                    $found = 1;
                    last;
                }
            }
        }
        ### if we still have not found one of the above man paths,
        ### just select the first one out of /etc/man.config
        unless ($found) {
            for my $line (@lines) {
                chomp $line;
                if ($line =~ m|^MANPATH\s+(\S+)|) {
                    $mpath = $1;
                    last;
                }
            }
        }
    }
    mkdir $mpath, 0755 unless -d $mpath;
    my $mfile = "${mpath}/${manpage}";
    print "[+] Installing $manpage man page as $mfile\n";
    copy $manpage, $mfile or die "[*] Could not copy $manpage to " .
        "$mfile: $!";
    chmod 0644, $mfile;
    print "[+] Compressing manpage $mfile\n";
    ### remove the old one so gzip doesn't prompt us
    unlink "${mfile}.gz" if -e "${mfile}.gz";
    system "$gzipCmd $mfile";
    return;
}

sub query_get_bleeding_snort() {
    my $ans = '';
    print "[+] Would you like to download the latest Snort rules from \n",
        "    http://$bleeding_snort_website/?\n";
    while ($ans ne 'y' && $ans ne 'n') {
        print "    ([y]/n)?  ";
        $ans = <STDIN>;
        return 1 if $ans eq "\n";
        chomp $ans;
    }
    if ($ans eq 'y') {
        return 1;
    }
    return 0;
}

sub check_commands() {
    my @path = qw(
        /bin
        /usr/bin
        /usr/local/bin
    );
    CMD: for my $cmd (keys %cmds) {
        unless (-x $cmds{$cmd}) {
            my $found = 0;
            PATH: for my $dir (@path) {
                if (-x "${dir}/${cmd}") {
                    $cmds{$cmd} = "${dir}/${cmd}";
                    $found = 1;
                    last PATH;
                }
            }
            unless ($found) {
                die "[*] Could not find $cmd, edit the ",
                    "config section of install.pl";
            }
        }
    }

    return;
}

sub query_preserve_config() {
    my $ans = '';
    while ($ans ne 'y' && $ans ne 'n') {
        print "[+] Would you like to preserve the config from the\n",
            '    existing fwsnort installation ([y]/n)?  ';
        $ans = <STDIN>;
        return 1 if $ans eq "\n";
        chomp $ans;
    }
    if ($ans eq 'y') {
        return 1;
    }
    return 0;
}

sub preserve_config() {
    my $file = 'fwsnort.conf';
    open C, "< $file" or die "[*] Could not open $file: $!";
    my @new_lines = <C>;
    close C;

    open CO, "< ${fwsnort_dir}/$file" or die "[*] Could not open ",
        "${fwsnort_dir}/$file: $!";
    my @orig_lines = <CO>;
    close CO;

    print "[+] Preserving existing config: ${fwsnort_dir}/$file\n";
    ### write to a tmp file and then move.
    my $printed_intf_warning = 0;
    open CONF, "> ${fwsnort_dir}/${file}.new" or die "[*] Could not open ",
        "${fwsnort_dir}/${file}.new: $!";
    for my $new_line (@new_lines) {
        if ($new_line =~ /^\s*#/) {
            print CONF $new_line;
        } elsif ($new_line =~ /^\s*(\S+)/) {
            my $var = $1;
            my $found = 0;
            for my $orig_line (@orig_lines) {
                if ($orig_line =~ /^\s*\S+INTF\s/) {
                    ### interfaces are no longer used!
                    unless ($printed_intf_warning) {
                        print "    NOTE: Interfaces are no longer used as of the ",
                        "0.8.0 release;\n    removing $var\n";
                        $printed_intf_warning = 1;
                    }
                }
                if ($orig_line =~ /^\s*$var\s/
                        and $orig_line !~ /INTF/) {
                    print CONF $orig_line;
                    $found = 1;
                    last;
                }
            }
            unless ($found) {
                print CONF $new_line;
            }
        } else {
            print CONF $new_line;
        }
    }
    close CONF;
    move "${fwsnort_dir}/${file}.new", "${fwsnort_dir}/$file";
    return;
}

sub usage() {
    my $exit = shift;
    print <<_HELP_;
install.pl: [options]
    -f, --force-mod-install        - Install all perl modules regardless
                                     of whether they already installed on
                                     the system.
    -F, --Force-mod-regex <regex>  - Install perl module that matches a
                                     specific regular expression.
    -E, --Exclude-mod-regex <re>   - Exclude a perl module that matches this
                                     regular expression.
    -u, --uninstall   - uninstall fwsnort
    -h, --help        - print help and exit
_HELP_
    exit $exit;
}
